Start-Process: Correct Unspecified WorkingDirectory Resolution

NoMoreFood opened the pull request:
# PR Summary

Modified Start-Process such that an unspecified working directory is correctly treated as literal when being resolved.

Fix #11940

## PR Context

Problem was discovered when investigating an error that resulted from executing Start-Process in a directory that contained wildcard characters. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

daxian-dbw commented:
Although it might be a grey area, this is a breaking change as it changes how the arguments passed to `WorkingDirectory`, `RedirectStandardError`, `RedirectStandardInput`, and `RedirectStandardOutput` get resolved. Need the committee to weigh in.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, we propose that instead of always treating it as a literal, we would ask to check if the path exists and if so, use it, otherwise fallback to the current behavior to retain existing behavior

vexx32 commented:
That makes sense for working directory, I would think... Don't think that makes a lot of sense for the redirection paths @SteveL-MSFT.

NoMoreFood commented:
@vexx32 Agreed.  In fact, when I force the command to match multiple matches for output redirection, the command says: `Start-Process: Cannot perform operation because the path resolved to more than one file. This command cannot operate on multiple files.`

@SteveL-MSFT Certainly can't say that makes a lot of sense to me, but I can adjust the code if that's the verdict.  Should the documentation be updated to say it "Sort Of" supports wildcards?  Thoughts on redirection parameters as @vexx32 mentions?

daxian-dbw commented:
> Redirection paths typically won't exist for most use cases

I agree that in most use cases the redirection files for standard output/error don't exist beforehand. But `-RedirectStandardInput` by definition requires an existing file. 
Even for `-RedirectStandardError` and `-RedirectStandardOutput`, it's hard to say the current wildcard behavior is not being used at all. I can imagine a CI system where a new log file is created every day in the format `ci-log-<date>.txt` and output/error of a process is supposed to be written to that log file. For this hypothetical scenario, it's natural to use a wildcard like `ci-log-*.txt` for those 2 parameters (only one file, which is created on that day, will be matched). I think that's why the committee decide to suggest a more conservative (less breaking) behavior.

> Working Directory by definition cannot be a useful wildcarded path, which can usually be assumed to exist.

One can create two directories named `[a]b` and `ab`. `[a]b` is a valid wildcard that matches `ab`. Again, I think the committee's suggestion is more conservative (less breaking).

iSazonov commented:
@daxian-dbw Based on your examples I can conclude that we should first make wildcard check and then literal.

SteveL-MSFT commented:
@iSazonov that's a valid point and would seem to be less breaking as any existing usage expecting wildcard matching would continue to work

NoMoreFood committed:
Correct Start-Process Unspecified WorkingDirectory Resolution

- Modified Start-Process such that an unspecified working directory is correctly treated as literal when being resolved.
- Added test for running Start-Process within a directory with with wildcard characters.

NoMoreFood commented:
@iSazonov @SteveL-MSFT @daxian-dbw @vexx32 It appears that PathUtils.ResolvePath() is not positioned well to do the sort of trial-and-error that's discussed.  Instead of biting off that task right now, what I've done adjusted this PR's scope to address the originally reported behavior that was causing Start-Process to fail when WorkingDirectory is not specified but the current directory is set to a path that contains wildcard-type characters.  (e.g. Set-Location 'C:\Path[]'; Start-Process cmd.exe was failing).  I _believe_ this change is non-breaking and should be non-controversial.

JamesWTruher commented:
I think there are additional problems here and definitely some holes in our testing. As the following shows:
```powershell
Describe "Should be able to set location to name with wild card" {
    BeforeAll {
        $testCases = @{ Name = 'foo[' },@{ Name = 'foo]' },@{ Name = 'foo[]' }, @{ Name = 'foo?' }, @{ Name = 'foo[?]' },
            @{ Name = 'foo[*]' },@{ Name = 'foo[*' },@{ Name = 'foo*]' }
    }

    It "Should be able to create and set the location to <Name>" -TestCases $testCases {
        param ( $Name )

        if ( $Name -match "\*" -and $IsWindows ) {
            Set-ItResult -Skipped -Because "'$Name' is not supported as directory name on Windows"
            return
        }
        
        $item = New-Item -Type Directory -Path ${TESTDRIVE}/$Name
        $item.Name | Should -Be $Name
        
        $i = Get-Item -LiteralPath ${TESTDRIVE}/$Name
        $i.Name | Should -Be $Name

        $location = Get-Location
        try {
            Set-Location -LiteralPath ${TESTDRIVE}/$Name
            $sName = [System.IO.Path]::GetFileName((Get-Location ).Path)
            $sName | Should -Be $Name
        }
        catch {
            throw "Set-Location $Name Failed"
        }
        finally { 
            Set-Location -LiteralPath $location.Path
        }
    }
}
```
when run on my Mac, I get the following result
```powershell
Describing Should be able to set location to name with wild card
  [+] Should be able to create and set the location to foo[ 36ms
  [+] Should be able to create and set the location to foo] 13ms
  [+] Should be able to create and set the location to foo[] 26ms
  [+] Should be able to create and set the location to foo? 28ms
  [+] Should be able to create and set the location to foo[?] 41ms
  [-] Should be able to create and set the location to foo[*] 56ms
    Expected strings to be the same, but they were different.
    Expected length: 6
    Actual length:   5
    Strings differ at index 4.
    Expected: 'foo[*]'
    But was:  'foo[]'
    18:         $i.Name | Should -Be $Name
    at <ScriptBlock>, /tmp/zap/tt.tests.ps1: line 18
  [-] Should be able to create and set the location to foo[* 17ms
    Expected strings to be the same, but they were different.
    String lengths are both 5.
    Strings differ at index 4.
    Expected: 'foo[*'
    But was:  'foo[]'
    18:         $i.Name | Should -Be $Name
    at <ScriptBlock>, /tmp/zap/tt.tests.ps1: line 18
  [-] Should be able to create and set the location to foo*] 20ms
    Expected strings to be the same, but they were different.
    String lengths are both 5.
    Strings differ at index 3.
    Expected: 'foo*]'
    But was:  'foo[]'
    18:         $i.Name | Should -Be $Name
    at <ScriptBlock>, /tmp/zap/tt.tests.ps1: line 18
```
The files present in `$TESTDRIVE` are:
```powershell
PS> ls $TESTDRIVE |ft -auto

    Directory: /private/tmp/ea03bb1e-4ae5-46b7-af36-6341e4015b20

UnixMode   User  Group  LastWriteTime Size Name
--------   ----  -----  ------------- ---- ----
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo?
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo[
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo[?]
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo[]
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo[*
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo[*]
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo]
drwxr-xr-x james wheel 4/8/2020 13:27   64 foo*]

```

NoMoreFood commented:
@JamesWTruher Yeah, I suspected that there are going to be these sort of wildcard processing issues I located in Start-Process in other cmdlets as well.  Unless you've already dove in, I'll look into the behavior you noted as a separate PR (unless you think it's somehow related to this specific fix for Start-Process).

NoMoreFood commented:
@JamesWTruher Yeah, I suspected that there are going to be these sort of wildcard processing issues I located in Start-Process in other cmdlets as well.  Unless you've already dove in, I'll look into the behavior you noted as a separate PR (unless you think it's somehow related to this specific fix for Start-Process).

Note, for Windows your match should be more like `$Name -match '[<>:"/\|?*]' -and $IsWindows` since none of these characters are supported.  When these illegal characters are eliminated, it passes on Windows but only the three test cases are executed.  I'll try on MacOS later.

iSazonov commented:
The more I think about wildcards in the scenario the less I like it. I'd prefer literal paths - it is unnatural behavior to expand wildcards for these paths. I can’t imagine that someone selects a working directory by wildcard, and especially a redirect file. It is unreliable and unpredictable.

NoMoreFood commented:
@JamesWTruher Reproduced your Mac results.  Set-Location -LiteralPath definitely doesn't handle processing '*' characters very well whereas the native shells do just fine.  I do believe this is a separate issue though from the one this PR addresses though.  Will look at the underlying code tomorrow to verify...

NoMoreFood commented:
@JamesWTruher I confirmed the behavior you reported is a different issue and is related to the code called within Set-Location that's responsible for case normalization.  I have a potential fix and will submit that separately.  Please proceed with adjudicating _this_ PR as-is.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -1896,7 +1896,7 @@ protected override void BeginProcessing()
             {
                 // Working Directory not specified -> Assign Current Path.
-                startInfo.WorkingDirectory = ResolveFilePath(this.SessionState.Path.CurrentFileSystemLocation.Path);
+                startInfo.WorkingDirectory = PathUtils.ResolveFilePath(this.SessionState.Path.CurrentFileSystemLocation.Path, this, true);
```
Can you use a named arg here so we know what `true` means?

NoMoreFood replied:
I can make that change, sure.

NoMoreFood replied:
Done

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, we agree that an implicit `WorkingDirectory` should be literal and never glob, so the targeted fix is fine for this PR (not a comment on the code itself but the intent).

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -1896,7 +1896,7 @@ protected override void BeginProcessing()
             {
                 // Working Directory not specified -> Assign Current Path.
-                startInfo.WorkingDirectory = ResolveFilePath(this.SessionState.Path.CurrentFileSystemLocation.Path);
+                startInfo.WorkingDirectory = PathUtils.ResolveFilePath(this.SessionState.Path.CurrentFileSystemLocation.Path, this, true);
```
this should probably be done in the helper function where *no* globbing should be done. I would suggest that `ResolveFilePath` (line 2989) should probably be calling the literalpath overload.
`string filepath = PathUtils.ResolveFilePath(path, this, true)`
I don't think _any_ of those files should be globbed

NoMoreFood replied:
@JamesWTruher I completely agree and that was the original version of this change (I force pushed overwrote the commit history, sorry).  The the committee determined that would be a breaking change since [presumably] somebody might have gone out of their way to escape those paths in the past. The recommended approach was globbing first and then, if no match was was found, doing a literal check.  Given my lack of enthusiasm for that approach as well as the fact there was no good helper function suited to doing that (that I know of), I scaled back the change.

NoMoreFood committed:
Correct Start-Process Unspecified WorkingDirectory Resolution - Review Changes

- Added named argument to ResolveFilePath() call for clarity.

iSazonov commented:
@daxian-dbw I think the PR is ready to merge.

daxian-dbw committed:
Fix the indentation

iSazonov commented:
@NoMoreFood Thanks for your contribution!
